validphys2/src/validphys/photon/compute.py

```diff
@@ -0,0 +1,265 @@
+"""Script that calls fiatlux to add the photon PDF."""
+import lhapdf
+
+import numpy as np
+
+from . import structure_functions as sf
+from scipy.interpolate import interp1d
+from scipy.integrate import trapezoid
+
+from eko.io import EKO
+from eko.io.manipulate import xgrid_reshape
+from eko.interpolation import XGrid
+
+from n3fit.io.writer import XGRID
+
+import fiatlux
+
+import yaml
+from os import remove
+import time
+
+class Photon:
+    def __init__(self, theoryid, fiatlux_runcard, replicas_id):
+        self.theory = theoryid.get_description()
+        self.fiatlux_runcard = fiatlux_runcard
+        self.replicas_id = replicas_id
+        self.q_in2 = 100**2
+
+        # parameters for the alphaem running
+        self.alpha_em_ref = self.theory["alphaqed"]
+        self.qref = self.theory["Qref"]
+        # TODO : maybe they shoud be kDIS instead of k, but usually they are the same
+        self.thresh_c = self.theory["kcThr"] * self.theory["mc"]
+        self.thresh_b = self.theory["kbThr"] * self.theory["mb"]
+        self.thresh_t = self.theory["ktThr"] * self.theory["mt"]
+        if self.theory["MaxNfAs"] <= 5 :
+            self.thresh_t = np.inf
+        if self.theory["MaxNfAs"] <= 4 :
+            self.thresh_b = np.inf
+        if self.theory["MaxNfAs"] <= 3 :
+            self.thresh_c = np.inf
+        self.set_betas()
+        self.set_thresholds_alpha_em()
+
+        # structure functions
+        self.qcd_pdfs = [lhapdf.mkPDF(fiatlux_runcard["pdf_name"], id) for id in replicas_id]
+
+        # TODO : maybe find a different name for fiatlux_dis_F2
+        path_to_F2 = theoryid.path / "fastkernel/fiatlux_dis_F2.pineappl.lz4"
+        path_to_FL = theoryid.path / "fastkernel/fiatlux_dis_FL.pineappl.lz4"
+        f2 = [sf.StructureFunction(path_to_F2, pdfs) for pdfs in self.qcd_pdfs]
+        fl = [sf.StructureFunction(path_to_FL, pdfs) for pdfs in self.qcd_pdfs]
+        f2lo = [sf.F2LO(pdfs, self.theory) for pdfs in self.qcd_pdfs]
+
+        self.path_to_eko_photon = theoryid.path / "eko_photon.tar"
+
+        # set fiatlux
+        ff = open('fiatlux_runcard.yml', 'w+')
+        yaml.dump(self.fiatlux_runcard, ff)
+        self.lux = [fiatlux.FiatLux('fiatlux_runcard.yml') for i in range(len(replicas_id))]
+        remove('fiatlux_runcard.yml')
+        # we have a dict but fiatlux wants a yaml file
+        # TODO : remove this dirty trick
```

Comment:
The problem is that once we add the `fiatlux` parameters to the runcard they arrive to this module as a dictionary, while `fiatlux` wants a path to a yaml file. We could even have a separate runcard for fiatlux and pass the path to this second runcard to this function, however in this case we have to deal with two runcards instead of only one

Reply:
Can it not be modified to take a dict as an input? It is not like the authors are unknown or external.

Reply:
I'm not sure since we are using the wrapper of a C++ code, but I will ask stefano C

Reply:
We had a look together with @niclaurenti, and `fiatlux` at the moment relies on a global `Settings` instance (a singleton), that contains a private map attribute, `_config`, and it is reading values always from there.
This `Settings` object has no setter but one, `load()`, that takes a string and interprets it as a file path.

Of course, we could add another function to the class, a constructor that directly creates an instance from a suitable map, and expose it to Python.
This would involve implementing it in C code, and update the Python bindings, + re-releasing and going through Conda once more. Everything doable, but it would take time and effort (even more considering that the author of this PR is not a `fiatlux` maintainer, and not used to it).

Reply:
My practical proposal is that, if we believe to be relevant enough, we can open an issue in `fiatlux` and postpone this improvement to later on, to better focus on improvements that are vp-related.

Reply:
The implementation here should use a temporary file to avoid name collisions. Ideally the file would be written to the vp output folder rather than the current directory. For all we know, the user has no permission to write there. Using the default for tempfile (`/tmp`) can be an option for this small file. There are a bunch of examples in the vp code.

That said, fixing this at the actual source (pressing "merge" on the automated pr of conda forge) would be appreciated and also easier to do robustly.

Reply:
> Using the default for tempfile (`/tmp`) can be an option for this small file.

I'd go for this, and it should be sufficiently simple. Thanks for the suggestion.

> That said, fixing this at the actual source (pressing "merge" on the automated pr of conda forge) would be appreciated and also easier to do robustly.

The person who should press "merge" is not the author of this PR, and sometimes things might break in other ways (it did previously).
I'm not suggesting to not do it, just not to consider blocking for this PR.

Reply:
FWIW the reason I tend to prefer not using `/tmp` is that clusters can mount it in interesting ways, requiring network transfers to some old drive. In general writing tempfiles in the place specified for that.

Reply:
> FWIW the reason I tend to prefer not using `/tmp` is that clusters can mount it in interesting ways, requiring network transfers to some old drive. In general writing tempfiles in the place specified for that.

I actually accepted that, thinking about the same issue, but lacking some cluster statistics since a while.

In principle, there is no safe place on every cluster, since most jobs are run from users' homes, and those are almost always not local to the node.

I considered doing it in memory, but then I'm not sure how to pass it from Python to C without passing through the system (if I should mount a memory file system to pass the file, I'd spend better my time fixing the issue in Fiatlux).